### PR TITLE
The Required Hirearchy added for the issue 49 in the models2

### PR DESCRIPTION
**Pull request by Milan933-coder**
# Neural-LAM New Hierarchy (PR README)

This document summarizes the class hierarchy refactor introduced in `models2`, with a side-by-side view of the old and new architecture.

## Old Hierarchy (`neural_lam/models`)

```text
pytorch_lightning.LightningModule
└── ARModel
    └── BaseGraphModel
        ├── GraphLAM
        └── BaseHiGraphModel
            ├── HiLAM
            └── HiLAMParallel
```

### Characteristics

- `ARModel` combined training loop logic and autoregressive unroll logic.
- Graph-specific prediction (`predict_step`) was inherited through the same training class tree.
- Probabilistic ensemble generation and deterministic training logic were not separated into dedicated modules.

## New Hierarchy (`neural_lam/models2`)

```text
torch.nn.Module
└── StepPredictor
    └── BaseGraphModel
        ├── GraphLAM
        └── BaseHiGraphModel
            ├── HiLAM
            └── HiLAMParallel

torch.nn.Module
└── Forecaster
    └── ARForecaster
        └── ARForecastSampler

pytorch_lightning.LightningModule
└── ForecasterModule
    └── EnsembleForecasterModule
```

### Composition in `models2`

- `ARForecaster` owns autoregressive rollout and calls `step_predictor.predict_step(...)`.
- `ARForecastSampler` extends `ARForecaster` and generates batched ensemble trajectories.
- `ForecasterModule` wraps any `Forecaster` for train/val/test flow.
- `EnsembleForecasterModule` extends `ForecasterModule` and switches training/validation scoring to CRPS for ensembles.

## What Changed Structurally

- **Decoupled one-step neural prediction from rollout strategy**:
  - `StepPredictor` handles single-step state transition logic.
  - `Forecaster`/`ARForecaster` handles temporal unrolling.
- **Decoupled rollout from Lightning training orchestration**:
  - `ForecasterModule` handles optimization, logging, and evaluation loops.
- **Added probabilistic ensemble path**:
  - `ARForecastSampler` for stochastic ensemble generation.
  - `EnsembleForecasterModule` for CRPS-based scoring.

## Old-to-New Responsibility Mapping

| Old (`models`) | New (`models2`) | Responsibility |
|---|---|---|
| `ARModel.predict_step` (abstract) | `StepPredictor.predict_step` | One-step model update |
| `ARModel.unroll_prediction` | `ARForecaster.forward` | AR rollout over time |
| `ARModel.common_step` | `ForecasterModule.common_step` | Batch unpack + forecast call |
| `ARModel.training_step` | `ForecasterModule.training_step` / `EnsembleForecasterModule.training_step` | Loss calculation + logging |
| N/A (dedicated ensemble sampler) | `ARForecastSampler` | Ensemble member generation |

## Files Added/Updated in `models2`

- `step_predictor.py`
- `base_graph_model.py` (now based on `StepPredictor`)
- `base_hi_graph_model.py`
- `forecaster.py`
- `ar_forecaster.py`
- `ar_forecast_sampler.py` (+ compatibility re-export in `ar_forecast_sample.py`)
- `forecaster_module.py`
- `ensemble_forecaster_module.py`

## Metrics Update

- Added `crps_loss` in `neural_lam/metrics.py` for sample-based ensemble CRPS evaluation.

## Development Note (In Progress)

- Current parity testing uses `noise_std=0.0` in `tests/test_models2_parity.py` to make CRPS comparisons deterministic across repeated forward passes.
- This is a temporary testing choice only; ensemble sampling is intended to run with non-zero noise (`noise_std > 0`) and this part is still under active development.

### CODE REVIEW

**Comment by joeloskarsson** [comment]
This is not a good way to propose these changes. You are almost refactoring the whole repo without anchoring the many design choices involved within the community. This is not how we do open source dev.

This should first of all build on #208, and then the changes and additions are best separated so one thing is added per PR.

**Comment by Milan933-coder** [comment]
Thanks for the feedback! I apologize for the massive PR—I am still getting used to the workflow for this repository. I created the separate models2 folder because I didn't want to break the existing code while you reviewed the changes for issue #49, but I understand now that this isn't the standard way to propose changes.
 I'll wait until #208 is merged, and then I will add the necessary changes as per your request accordingly. Let me know if that works!

**Comment by Milan933-coder** [comment]
Thanks for the honest feedback. I am still learning the ropes of standard open-source workflows, and I really appreciate you pointing me in the right direction.

**Comment by Milan933-coder** [comment]
Sir,I Have My Submitted My Propsal for adding EnsembleForecasterModlue, ArForecasterSampler, CNNPRedictor and Merging the prob_model_global,prob_model_lam with the main branch
